### PR TITLE
Allow sorting of group headers within any given hierarchy

### PR DESCRIPTION
**Pull request by awseward**
### Changes Made

Initially, this was implemented to order group headers within a given hierarchy according to the order they were encountered in (a decision made in order to respect the intent of [the original ListGroupable](https://github.com/procore/react-widgets/blob/bc933b37f1756979a784b6b806e6f3dc43670086/src/ListGroupable.jsx#L191-L193)).

This change allows us to sort the headings per hierarchy level.
### Example

Here's an example, in case that description above doesn't quite paint a clear picture.
##### Before

[The requirement for this](https://github.com/procore/wrench/pull/1283) was that the `Budgeted/Non-Budgeted` grouping needed to be sorted such that `Budgeted` always came first, but with the initial implementation of n-tier grouping, you're sort of just at the mercy of which groups were found first in the flat `data` array that the groups are constructed from.

This means that, depending on the data and the query filter, sometimes `Budgeted` would be first, but other times `Non-Budgeted` would be first. One such case of `Non-Budgeted` coming first is shown below.
screen shot 2016-05-04 at 11 21 18
##### After

With this change, the caller can enforce an ordering for any given tier of group headers as shown below. (same data, same query filter). Notice how now the `Budgeted` group is shown before the `Non-Budgeted` group, as per the requirement.
screen shot 2016-05-04 at 11 22 47

### CODE REVIEW

File: src/util/objectTraversal.js [hunk 1]
```diff
@@ -24,7 +21,9 @@ export function depthFirst(currentNode, getChildren, onInternal, onLeaf, state)
     });
   }
 
-  return getChildren(currentNode).reduce(
+  const getCurrentChildren = getChildren[0] || (x => Object.keys(s));
```
Comment (remyyounes) [inline src/util/objectTraversal.js]: Is this supposed to be `(x => Object.keys(s))` or `(x => Object.keys(x))`

**Comment by pmkhougaz** [comment]
+1 pending Remy's comments

File: src/util/objectTraversal.js (hunk 1 above)
Comment (awseward) [inline src/util/objectTraversal.js]: @remyyounes: Ah, good catch.

**Comment by yoyis3000** [comment]
+1

**Comment by zgross** [comment]
+1

**Comment by denzyl** [comment]
+1

**Comment by kespinola** [comment]
+1
Sensible, but counter requirement. It feels like we need a groupByRules. Procore want more than just groupBy... it wants total list control.

**Comment by thomasm-procore** [comment]
Passed testing.

**Pull request merged**